provers/sp1/driver/src/distributed/worker/pool.rs

```diff
@@ -0,0 +1,264 @@
+use std::{
+    collections::{BTreeMap, VecDeque},
+    sync::Arc,
+    time::Duration,
+};
+
+use raiko_lib::prover::WorkerError;
+use sp1_core::{air::PublicValues, utils::SP1CoreOpts};
+use tokio::sync::RwLock;
+
+use crate::{
+    sp1_specifics::{Challenger, Checkpoint, Commitments, PartialProofs, ShardsPublicValues},
+    RequestData, WorkerRequest, WorkerResponse, WorkerSocket,
+};
+
+pub struct WorkerPool {
+    workers: BTreeMap<usize, Arc<RwLock<WorkerSocket>>>,
+}
+
+impl WorkerPool {
+    pub async fn new() -> Result<Self, WorkerError> {
+        let workers = Self::spawn_workers().await?;
+
+        Ok(Self { workers })
+    }
+
+    pub async fn commit(
+        &mut self,
+        checkpoints: Vec<(Checkpoint, usize)>,
+        public_values: PublicValues<u32, u32>,
+        opts: SP1CoreOpts,
+    ) -> Result<(Commitments, Vec<ShardsPublicValues>), WorkerError> {
+        let requests = checkpoints
+            .into_iter()
+            .map(|(checkpoint, nb_checkpoints)| {
+                WorkerRequest::Commit(RequestData {
+                    checkpoint,
+                    nb_checkpoints,
+                    public_values,
+                    shard_batch_size: opts.shard_batch_size,
+                    shard_size: opts.shard_size,
+                })
+            })
+            .collect();
+
+        let commitments_response = self.distribute_work(requests).await?;
+
+        let mut commitments_vec = Vec::new();
+        let mut shards_public_values_vec = Vec::new();
+
+        for response in commitments_response {
+            if let WorkerResponse::Commitment {
+                commitments,
+                shards_public_values,
+            } = response
+            {
+                commitments_vec.extend(commitments);
+                shards_public_values_vec.extend(shards_public_values);
+            } else {
+                return Err(WorkerError::InvalidResponse);
+            }
```

Comment:
Could be refactored into `let ... else` form. Not that important, it makes it into a guard clause which is more readable IMO.

Something like:
```suggestion
            let WorkerResponse::Commitment {
                commitments,
                shards_public_values,
            } = response else {
                return Err(WorkerError::InvalidResponse);
            };
            
            commitments_vec.extend(commitments);
            shards_public_values_vec.extend(shards_public_values);
```

Reply:
This is indeed more readable :)